rllib/utils/checkpoints.py

```diff
@@ -29,15 +37,287 @@
 # 1.1: Same as 1.0, but has a new "format" field in the rllib_checkpoint.json file
 # indicating, whether the checkpoint is `cloudpickle` (default) or `msgpack`.
 
-# 1.2: Introduces the checkpoint for the new Learner API if the Learner api is enabled.
+# 1.2: Introduces the checkpoint for the new Learner API if the Learner API is enabled.
+
+# 2.0: Introduces the Checkpointable API for all components on the new API stack
+# (if the Learner-, RLModule, EnvRunner, and ConnectorV2 APIs are enabled).
 
 CHECKPOINT_VERSION = version.Version("1.1")
 CHECKPOINT_VERSION_LEARNER = version.Version("1.2")
+CHECKPOINT_VERSION_LEARNER_AND_ENV_RUNNER = version.Version("2.0")
+
+
+@PublicAPI(stability="alpha")
+class Checkpointable(abc.ABC):
+    """Abstract base class for a component of RLlib that can be checkpointed to disk.
+
+    Subclasses must implement the following APIs:
+    - save_to_path()
+    - restore_from_path()
+    - from_checkpoint()
+    - get_state()
+    - set_state()
+    - ctor_args_and_kwargs()
+    - get_metadata()
+    - get_checkpointable_components()
+    """
+
+    # The subdirectory of this class (if it's a subcomponent of another Checkpointable).
+    # For example, if A is-a Checkpointable and contains an instance
+    # of class B (also a Checkpointable), then class B should set this to something like
+    # "component_B". When class A's `save_to_path([path])` is called, the state of
+    # class B can then be found in the dir: `path/component_B/`.
+    COMPONENT_DIR_NAME = ""
+
+    # The state file for the implementing class.
+    # This file contains any state information that does NOT belong to any subcomponent
+    # of the implementing class (which are `Checkpointable` themselves and thus should
+    # have their own state- and metadata files).
+    # After a `save_to_path([path])` this file can be found directly in: `path/`.
+    STATE_FILE_NAME = "state.pkl"
+
+    # The filename of the pickle file that contains the class information of the
+    # Checkpointable as well as all constructor args to be passed to such a class in
+    # order to construct a new instance.
+    CLASS_AND_CTOR_ARGS_FILE_NAME = "class_and_ctor_args.pkl"
+
+    # Subclasses may set this to their own metadata filename.
+    # The dict returned by self.get_metadata() is stored in this JSON file inside
+    # `COMPONENT_DIR_NAME/`.
+    METADATA_FILE_NAME = "metadata.json"
+
+    def save_to_path(
+        self,
+        path: Optional[Union[str, pathlib.Path]] = None,
```

Comment:
This does exclude paths to cloud storage like S3 I think. In the future we should provide this. When running with `tune` I guess `tune`  takes care of this b/c it syncs basically the driver files with the S3 bucket (or other object storage).

Reply:
Correct. Actually, for `Algorithm` I'm not sure yet, how we are going to make this an implementer of Checkpointable API as it might collide with `tune.Trainable` in exactly these functionalities.

Keep in mind that the subcomponents of RLlib (e.g. RLModule, Learner, etc..) - if used in isolation - will never be run in connection with Tune, so this should be fine. It would be super nice, though, to have Algorithm also abide to this new API, but we'll see. For now, it'll use the tune.Trainable's `save_checkpoint`, which takes care of S3/remote synching.